Eliminate build warnings from the main project

logiclrd opened the pull request:
This PR makes minor adjustments so that `rustdesk` builds without warnings.

Before:

```
C:\code\rustdesk>cargo build 2>&1 | find "warning"
warning: unused imports: `MOUSE_BUTTON_LEFT`, `MOUSE_TYPE_DOWN`, `MOUSE_TYPE_UP`, and `MOUSE_TYPE_WHEEL`
warning: use of deprecated function `hbb_common::base64::encode`: Use Engine::encode
warning: variable does not need to be mutable
warning: fields `x` and `y` are never read
warning: constant `MOUSE_MOVE_PROTECTION_TIMEOUT` is never used
warning: constant `MOUSE_ACTIVE_DISTANCE` is never used
warning: function `get_last_input_cursor_pos` is never used
warning: methods `read_empty_dirs`, `set_confirm_override_file`, and `rename_file` are never used
warning: function `query_online_states` is never used
warning: function `create_online_stream` is never used
warning: function `query_online_states_` is never used
warning: methods `http_request` and `get_http_status` are never used
warning: methods `into_string` and `save` are never used
warning: function `get_chatid_telegram` is never used
warning: static `CONTROLLING_SESSION_COUNT` is never used
warning: function `update_controlling_session_count` is never used
warning: function `http_request` is never used
warning: function `get_async_http_status` is never used
warning: function `has_valid_bot` is never used
warning: function `verify_bot` is never used
warning: enum `HbbHttpResponse` is never used
warning: struct `DownloadData` is never constructed
warning: struct `Downloader` is never constructed
warning: function `download_file` is never used
warning: function `do_download` is never used
warning: function `get_download_data` is never used
warning: function `cancel` is never used
warning: function `remove` is never used
warning: `rustdesk` (lib) generated 28 warnings (run `cargo fix --lib -p rustdesk` to apply 2 suggestions)
warning: unused import: `librustdesk::*`
warning: `rustdesk` (bin "service") generated 1 warning (run `cargo fix --bin "service"` to apply 1 suggestion)

C:\code\rustdesk>
```

After:

```
C:\code\rustdesk>cargo clean -p rustdesk
     Removed 401 files, 6.6GiB total

C:\code\rustdesk>cargo build
   Compiling rustdesk v1.4.3 (C:\code\rustdesk)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 40.47s

C:\code\rustdesk>
```

In this PR, each type of warning is addressed in a separate commit.

logiclrd committed:
Marked a number of blocks and crates with [allow(dead_code)] and, more generally, [allow(unused)], as static analysis apparently doesn't see the call sites.

logiclrd committed:
Updated use of hbb_common::base64 (republished from sodiumoxide) in hbbs_http/sync.rs to disable warning about it being deprecated.

logiclrd committed:
Removed unused import of librustdesk::* in src/service.rs.

logiclrd committed:
Updated ui_session_interface.rs to allow unused imports.

logiclrd committed:
Updated tray.rs to not mark event_loop as mutable.

logiclrd commented:
Huh. There's some code that _does_ require it to be mutable, but that code is only enabled for the `macos` target.

logiclrd committed:
Updated tray.rs to conditionally declare event_loop's mutability based on whether the code that requires it to be mutable is enabled.

rustdesk commented on src/hbbs_http/sync.rs:
```diff
@@ -184,7 +184,10 @@ async fn start_hbbs_sync_async() {
+                        {
+                            #![allow(deprecated)]
+                            hash = hbb_common::base64::encode(&res[..]);
+                        }
```
should change to encode64 in src/common.rs

logiclrd replied:
Ah, nice, centralization of the override. 🙂

logiclrd committed:
Switched call to hbb_common::base64::encode in sync.rs to use crate::encode64 instead.

logiclrd committed:
Updated the macos main() function in service.rs to locally import librustdesk::*.

logiclrd committed:
Reworked the macos main method to import librustdesk itself and then qualify references with it.

logiclrd commented:
There are some OS X-specific warnings that I have no idea what to do with:

```
Warning: warning: rustdesk@1.4.3: src/platform/macos.mm:96:18: warning: 'AuthorizationExecuteWithPrivileges' is deprecated: first deprecated in macOS 10.7 [-Wdeprecated-declarations]
warning: rustdesk@1.4.3:         status = AuthorizationExecuteWithPrivileges(authRef, process, kAuthorizationFlagDefaults, args, &pipe);
warning: rustdesk@1.4.3:                  ^
warning: rustdesk@1.4.3: /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Security.framework/Headers/Authorization.h:427:10: note: 'AuthorizationExecuteWithPrivileges' has been explicitly marked deprecated here
warning: rustdesk@1.4.3: OSStatus AuthorizationExecuteWithPrivileges(AuthorizationRef authorization,
warning: rustdesk@1.4.3:          ^
Warning: warning: rustdesk@1.4.3: src/platform/macos.mm:134:30: warning: 'CGDisplayModeCopyPixelEncoding' is deprecated: first deprecated in macOS 10.11 - No longer supported [-Wdeprecated-declarations]
warning: rustdesk@1.4.3:         CFStringRef pixelEncoding = CGDisplayModeCopyPixelEncoding(mode);
warning: rustdesk@1.4.3:                                     ^
warning: rustdesk@1.4.3: /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGDirectDisplay.h:177:34: note: 'CGDisplayModeCopyPixelEncoding' has been explicitly marked deprecated here
warning: rustdesk@1.4.3: CG_EXTERN CFStringRef __nullable CGDisplayModeCopyPixelEncoding(
warning: rustdesk@1.4.3:                                  ^
```

logiclrd committed:
Eliminate warnings appearing on a Linux build:
- Marked MouseButton and ScrollDirection in server/uinput.rs as allow(unused).
- Removed the resolution field from RdpInputMouse, as nothing actually ever reads from it. (The constructor still takes it, as it derives instance data from its value.)
- Updated server/display_service.rs to only generate the is_capturer_mag_supported function and the IS_CAPTURER_MAGNIFIER_SUPPORTED lazy static on the "windows" configuration.
- Updated server/connection.rs to not import crate::clipboard_file::* for target_os "linux".
- Updated the conditional attribute of enable_file_transfer in server/connection.rs to match the condition applied to its use (target_os "windows" | feature "unix-file-copy-paste").
- Updated align_to_32 in server/audio_service.rs to not assign a value to buf that it then immediately overwrites.
- Removed unnecessary assignment to local "conn" in a lambda within sync_and_watch_config_dir in server.rs.
- Moved the is_opensuse() function in platform/linux.rs into the comment block wrapping function elevate, which is its only consumer.
- Updated platform/linux.rs to use whoami::fallible::hostname instead of the deprecated whoami::hostname.
- Updated platform/linux.rs to mark the AwakeHandle inside WakeLock with [allow(unused)], since nothing explicitly reads it; its purpose is simply its presence.
- Updated flutter_ffi.rs to conditionally import PathBuf for the platforms that have code explicitly using it, and to not import time::Duration at all. I think it is unnecessary??

logiclrd committed:
Updated flutter_ffi.rs to import std::time::Duration but only for target_os "windows" and "macos".

logiclrd committed:
Added a conditional import of unix_file_clip to server/connection.rs.

logiclrd committed:
Removed the explicit import of crate::unix_file_clip, which Copilot suggested but was wrong anyway, with increasing the scope of the import of crate::clipboard_file::* to include feature "unix-file-copy-paste".

rustdesk commented on src/server.rs:
```diff
@@ -703,7 +703,6 @@ async fn sync_and_watch_config_dir() {
                                 log::error!("sync config to root failed: {}", e);
                                 match crate::ipc::connect(1000, "_service").await {
                                     Ok(mut _conn) => {
-                                        conn = _conn;
```
Really? This creates a bug. That's why I hate refactory.

crate::ipc::connect creates a new connection.

logiclrd replied:
You should turn on `[deny(warnings)]` so that the code can be correct without warnings in the first place. This is a very common practice.

logiclrd replied:
Here's why I removed that line:

<img width="662" height="172" alt="image" src="https://github.com/user-attachments/assets/3af05828-63da-41fa-926e-e53af856e624" />

Is this just a bug in the analyzer, then?

rustdesk replied:
You are right. Here is a bug, the `break` should be  removed.

logiclrd commented on src/server/connection.rs:
```diff
@@ -3,7 +3,7 @@ use super::{input_service::*, *};
 #[cfg(not(any(target_os = "android", target_os = "ios")))]
 use crate::clipboard::{update_clipboard, ClipboardSide};
-#[cfg(any(target_os = "windows", target_os = "linux", target_os = "macos"))]
+#[cfg(any(target_os = "windows", target_os = "macos", feature = "unix-file-copy-paste"))]
```
I actually don't know if it'd work without including it for the "macos" target OS. I guess I'll try removing it and then revert if the "macos" CI fails (I have no other way to test it).

logiclrd replied:
Nice, build succeeded. :slightly_smiling_face:

logiclrd committed:
Removed "macos" target OS as one of the conditions for including crate::clipboard_file::* in connection.rs. This might break the OS X build, in which case I will undo this change.

logiclrd commented on src/server/display_service.rs:
```diff
@@ -22,8 +22,12 @@ struct ChangedResolution {
     static ref IS_CAPTURER_MAGNIFIER_SUPPORTED: bool = is_capturer_mag_supported();
+}
+
+lazy_static::lazy_static! {
```
AI hallucination. This isn't the conventional practice because it doesn't work at all.

```
error[E0573]: expected type, found static `IS_CAPTURER_MAGNIFIER_SUPPORTED`
  --> src/server/display_service.rs:25:1
   |
25 | / lazy_static::lazy_static! {
26 | |     #[cfg(windows)]
27 | |     static ref IS_CAPTURER_MAGNIFIER_SUPPORTED: bool = is_capturer_mag_supported();
28 | | }
   | |_^ not a type
   |
   = note: this error originates in the macro `__lazy_static_internal` which comes from the expansion of the macro `lazy_static::lazy_static` (in Nightly builds, run with -Z macro-backtrace for more info)
```

There are plenty of examples in the code that use multiple `lazy_static!` macro invocations with different conditions exactly like I wrote here. For instance, in `hbb_common`'s `config.rs`:

```
#[cfg(target_os = "macos")]
lazy_static::lazy_static! {
    pub static ref ORG: RwLock<String> = RwLock::new("com.carriez".to_owned());
}

type Size = (i32, i32, i32, i32);
type KeyPair = (Vec<u8>, Vec<u8>);

lazy_static::lazy_static! {
    static ref CONFIG: RwLock<Config> = RwLock::new(Config::load());
    static ref CONFIG2: RwLock<Config2> = RwLock::new(Config2::load());
    ...
```

And in `client.rs`:

```
lazy_static::lazy_static! {
    pub(crate) static ref CUR_SESSION_ID: RwLock<SessionID> = Default::default(); // For desktop only
    static ref GLOBAL_EVENT_STREAM: RwLock<HashMap<String, StreamSink<String>>> = Default::default(); // rust to dart event channel
}

#[cfg(target_os = "windows")]
lazy_static::lazy_static! {
    pub static ref TEXTURE_RGBA_RENDERER_PLUGIN: Result<Library, LibError> = load_plugin_in_app_path("texture_rgba_renderer_plugin.dll");
}

#[cfg(target_os = "linux")]
lazy_static::lazy_static! {
    pub static ref TEXTURE_RGBA_RENDERER_PLUGIN: Result<Library, LibError> = Library::open("libtexture_rgba_renderer_plugin.so");
}

#[cfg(target_os = "macos")]
lazy_static::lazy_static! {
    pub static ref TEXTURE_RGBA_RENDERER_PLUGIN: Result<Library, LibError> = Library::open_self();
}
```

logiclrd committed:
PR feedback: Replaced unwrap_or_else with unwrap_or_default in use of fallible::hostname() in platform/linux.rs, because the value the functor was yielding was the default anyway.

logiclrd committed:
Restored assignment of re-established connection to the loop variable in sync_and_watch_config_dir in server.rs, now that it is no longer unused because of ce7d794b.

logiclrd commented:
There's another unused symbol warning that has the "feel" of a bug to it. It might just legitimately be unused, but since the last time this happened it was actually a logic error in the code, I figured I'd bring it up.

Specifically, the `sync_and_watch_config_dir` function in `src/server.rs` has a retry loop, and it has a local variable `synced` that starts out `false` but gets set to `true` once it has successfully synchronized configuration from the root. It skips the downstream synchronization process if it is already `synced`. But, immediately after that, as far as I can see it enters an unconditional loop watching for local changes to push back up to the root and there is no way for it to get out of that loop. The outer loop is effectively no longer active at this point, which means that the value assigned to `synced` won't ever be checked. I _think_ that means that `synced` is actually completely redundant with the current structure and should simply be removed?

Pseudocode:

```
synced = false;

for i in retries
{
  sleep(interval);

  result = ipc::connect(..);

  if (result is Ok)
  {
    if (!synced)
    {
      sync downstream
      synced = true;
    }

    loop forever
    {
      sleep(interval);

      if (configuration has changed)
      {
        sync upstream
        reconnect if that failed
      }
    }
  }
}
```

I think this would do exactly the same thing without `synced`:

```
for i in retries
{
  sleep(interval);

  result = ipc::connect(..);

  if (result is Ok)
  {
    sync downstream

    loop forever
    {
      sleep(interval);

      if (configuration has changed)
      {
        sync upstream
        reconnect if that failed
      }
    }
  }
}
```

logiclrd committed:
Removed unused import of anyhow from src/server/wayland.rs.

workingjubilee commented on src/server/display_service.rs:
```diff
@@ -22,8 +22,12 @@ struct ChangedResolution {
+#[cfg(windows)]
 lazy_static::lazy_static! {
     static ref IS_CAPTURER_MAGNIFIER_SUPPORTED: bool = is_capturer_mag_supported();
+}
```
Could be replaced by https://doc.rust-lang.org/std/sync/struct.LazyLock.html

```suggestion
static IS_CAPTURER_MAGNIFIER_SUPPORTED: LazyLock<bool> = LazyLock::new(|| {
    scrap::CapturerMag::is_supported()
});
```

Then you can remove the extraneous "forwarding" function.

workingjubilee commented on src/tray.rs:
```diff
@@ -57,7 +57,10 @@ fn make_tray() -> hbb_common::ResultType<()> {
+    #[cfg(target_os = "macos")]
     let mut event_loop = EventLoopBuilder::new().build();
+    #[cfg(not(target_os = "macos"))]
+    let event_loop = EventLoopBuilder::new().build();
```
This can use a block that evaluates to the `event_loop` after its mutation so that arbitrary additional mutation doesn't occur:
```suggestion
    #[cfg(target_os = "macos")]
    let event_loop = {
        use tao::platform::macos::{EventLoopExtMacOS, ActivationPolicy};
        let mut event_loop = EventLoop::new();
        event_loop.set_activation_policy(ActivationPolicy::Accessory);
        event_loop
    };
    #[cfg(not(target_os = "macos"))]
    let event_loop = EventLoop::new();
```

logiclrd replied:
I think this would work, right?

```rust
    let event_loop = {
        #[allow(unused_mut)]
        let mut event_loop = EventLoopBuilder::new().build();

        #[cfg(target_os = "macos")]
        {
            use tao::platform::macos::EventLoopExtMacOS;
            event_loop.set_activation_policy(tao::platform::macos::ActivationPolicy::Accessory);
        }

        event_loop
    };
```

Looks like it compiles, and it eliminates a lot of the complexity/repetitiveness. 🙂

workingjubilee replied:
Yep, that works too. Just different ways of saying "let's try to *not* mutate after this point, mmk?"

logiclrd committed:
Updated IS_CAPTURER_MAGNIFIER_SUPPORTED from lazy_static to LazyLock<bool> in display_service.rs.

logiclrd committed:
Restructured the initialization of event_loop to constrain the mutability of event_loop to a tight initialization scope, and consolidated the per-platform definitions to a single definition.

workingjubilee commented on src/tray.rs:
Yes.

logiclrd replied:
> Apparently I cannot resolve this, but

As far as I can see, the lines are needed (at least, removing them would be a semantic change), but they can be moved into the constrained scope of mutability -- as is the case in the latest commit -- so that `event_loop` doesn't need to be mutable in the broader scope.

```
     let event_loop = {
         #[allow(unused_mut)]
         let mut event_loop = EventLoopBuilder::new().build();
 
>        #[cfg(target_os = "macos")]
>        {
>            use tao::platform::macos::EventLoopExtMacOS;
>            event_loop.set_activation_policy(tao::platform::macos::ActivationPolicy::Accessory);
>        }
 
         event_loop
     };
```
